prereise/gather/demanddata/bldg_electrification/puma_hp_estimator.py

```diff
@@ -0,0 +1,373 @@
+                stats[
+                    [
+                        "frac_elec_htg",
+                        "frac_elec_htg_sq",
+                    ]
+                ]
+            )
+        ),
+    ).fit()
+    hp_fit_df_low_elec = pd.Series(
+        data=[
+            lm_hp_low.params[0],
+            lm_hp_low.params[1],
+            lm_hp_low.params[2],
+        ],
+        index=[
+            "const",
+            "s_frac_elec_htg",
+            "s_frac_elec_htg_sq",
+        ],
+    )
+    fit_stats_low_elec = pd.Series(
+        data=[
+            lm_hp_low.bse[0],
+            lm_hp_low.bse[1],
+            lm_hp_low.bse[2],
+            lm_hp_low.pvalues[1],
+            lm_hp_low.pvalues[2],
+            lm_hp_low.rsquared,
+        ],
+        index=[
+            "stderr_const",
+            "stderr_frac_elec_htg",
+            "stderr_frac_elec_htg_sq",
+            "pvalue_frac_elec_htg",
+            "pvalue_frac_elec_htg_sq",
+            "rsquared",
+        ],
+    )
+
+    return hp_fit_df, fit_stats, hp_fit_df_low_elec, fit_stats_low_elec
+
+
+def estimate_puma_hp_penetration(hp_fit_df, hp_fit_df_low_elec, puma_data_metro):
+    """Compute current heat pump penetration for pumas, and then scale the values to align with regional inputs from
+         RECS and CBECS surveys
+
+    :param pandas.DataFrame hp_fit_df: slope and intercept of heat pump penetration regression model, output of
+        hp_penetration_fit()
+    :param pandas.DataFrame hp_fit_df_low_elec: slope and intercept of heat pump penetration regression model for
+        low electrified locations, output of hp_penetration_fit()
+    :param pandas.DataFrame puma_data_metro: puma data within metropolitan areas, output of puma_county_shp_overlay()
+
+    :return: (*pandas.DataFrame*) hp_puma_df: current heat pump penetration estimations for pumas that has been scaled
+        based on RECS and CBECS surveys
+    """
+
+    puma_fuel = const.puma_fuel
+    puma_data = const.puma_data
+    natgas_price = pd.read_csv(
+        os.path.join(
+            os.path.dirname(__file__), "data", "state_ng_residential_price_2019.csv"
+        ),
+        index_col=0,
+    )
+    puma_hp_df = pd.DataFrame(
+        {
+            "state": puma_fuel["state"],
+            "frac_elec_htg": puma_fuel["hh_elec"] / puma_fuel["hh_total"],
+            "log_hdd": np.log(puma_data["hdd65_normals"]),
+            "natgas_price": puma_fuel.apply(
+                lambda x: natgas_price.loc[x.loc["state"], "$/kWh"], axis=1
+            ),
+            "hh_total": puma_fuel["hh_total"],
+        },
+        index=puma_data.index,
+    )
+    puma_hp_df["frac_hp_fitted"] = puma_hp_df.apply(
+        lambda x: (
+            x["frac_elec_htg"] * hp_fit_df["s_frac_elec_htg"]
+            + x["log_hdd"] * hp_fit_df["s_log_hdd"]
+            + x["natgas_price"] * hp_fit_df["s_natgas_price"]
+            + hp_fit_df["const"]
+        )
+        if x["frac_elec_htg"] > 0.27
```

Comment:
Let's add some simple doc strings for this threshold which splits high variance model and low variance model. More detailed discussions should be added in the corresponding section of the project documentation.

Reply:
Sure